app/src/main/java/com/mapbox/maps/testapp/examples/LargeGeojsonPerformanceActivity.kt

```diff
@@ -0,0 +1,163 @@
+ */
+class LargeGeojsonPerformanceActivity : AppCompatActivity() {
+
+  private lateinit var routePoints: FeatureCollection
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_simple_map)
+
+    routePoints = FeatureCollection.fromFeature(
+      Feature.fromGeometry(
+        LineString.fromPolyline(
+          DirectionsResponse.fromJson(
+            AnnotationUtils.loadStringFromAssets(
+              this@LargeGeojsonPerformanceActivity, LARGE_GEOJSON_ASSET_NAME
+            )
+          ).routes()[0].geometry()!!,
+          Constants.PRECISION_6
+        )
+      )
+    )
+
+    mapView.getMapboxMap()
+      .apply {
+        setCamera(
+          CameraOptions.Builder()
+            .center(Point.fromLngLat(LONGITUDE, LATITUDE))
+            .zoom(START_ZOOM)
+            .build()
+        )
+        // start an animation that uses UI thread to update map camera
+        flyTo(
+          CameraOptions.Builder()
+            .zoom(END_ZOOM)
+            .build(),
+          MapAnimationOptions.mapAnimationOptions {
+            duration(ANIMATION_DURATION_MS)
+          }
+        )
+        // start loading style with multiple very heavy geojson's
+        loadStyle(
+          style(Style.MAPBOX_STREETS) {
+            for (i in 0 until LARGE_SOURCE_COUNT) {
+              +geoJsonSource("${SOURCE}_$i") {
+                featureCollection(routePoints)
+              }
+              +lineLayer("${LAYER}_$i", "${SOURCE}_$i") {
+                lineColor("blue")
+                lineOffset(5.0 * i)
+              }
+            }
+            // add an icon that uses very small geojson source
+            +image("icon") {
+              bitmap(BitmapFactory.decodeResource(resources, R.drawable.blue_marker_view))
+            }
+            +geoJsonSource("${SOURCE}_marker") {
+              geometry(Point.fromLngLat(LONGITUDE, LATITUDE))
+            }
+            +symbolLayer("${LAYER}_marker", "${SOURCE}_marker") {
+              iconImage("icon")
+              iconAnchor(IconAnchor.BOTTOM)
+            }
+          }
+        ) {
+          // also add couple of additional large geojson's when style is fully loaded
+          loadAdditionalGeoJsonAfter(it)
+        }
+      }
+  }
+
+  private fun loadAdditionalGeoJsonAfter(style: Style) {
+    // async method with adding layer in callback when geojson is completely parsed
+    geoJsonSource(
+      id = "${SOURCE}_$LARGE_SOURCE_COUNT",
+      block = {
+        featureCollection(routePoints)
+      },
+      onGeoJsonParsed = {
+        style.addSource(it)
+        style.addLayer(
+          lineLayer("${LAYER}_$LARGE_SOURCE_COUNT", "${SOURCE}_$LARGE_SOURCE_COUNT") {
+            lineColor("green")
+            lineOffset(5.0 * LARGE_SOURCE_COUNT)
+          }
+        )
+      }
+    )
+    // async method with adding layer instantly without waiting for geojson to actually parse
```

Comment:
As far as I remember, without specifying the onDataParsed closure, the geojson parsing is processed in sync manner?

Reply:
No, because we're using DSL function here. Sync mode will be if:
```
val source = style.getSourceAs(id)
source.feature(someFeature) // this will be sync
source.feature(someFeature, {}) // this will be async
```

Reply:
I think below code(`style.addSource()`) doesn't use StyleExtension, so it won't be async?

Reply:
Adding source of course is sync but parsing is still async. Whenever `GeoJsonSource.Builder` is used - we will always perform geojson parsing using worker thread.
The only way to parse featureCollection etc on main thread is to call method `featureCollection` directly on `GeoJsonSource` object.

Reply:
So in this case we immediately add source to style (with no data), connect it with layer and when data is parsed in async mode we update to source and data becomes visible. Main difference with not using `onGeoJsonParsed` here is that we do not know exact moment of time when actual data becomes visualised, it will simply become visualised ASAP.

Reply:
ok, now I remember.. We probably need to provide some guide on the GeoJson parsing, since this is getting complex 😄